### PR TITLE
sched: better support of HistogramVec in scheduler performance test

### LINKED ISSUE #99182: sched: update metric type to HistogramVector in scheduler performance test
The new scheduler performance test was introduced back in 1.18, back then, the metric it supported are all `Histogram`. However, along with #92202, some metrics are updated to be `HistogramVector`, and hence it's no longer appropriate to return the first Histogram:

https://github.com/kubernetes/kubernetes/blob/5bb03db026e796f6c407f796dbf7ddcae0b0eb1f/staging/src/k8s.io/component-base/metrics/testutil/metrics.go#L206-L210

However, it's very common (for instance Preemption) that the metrics may hold multiple histograms:

```
metricName = scheduler_e2e_scheduling_duration_seconds
- metric.Label = [name:"profile" value:"default-scheduler"  name:"result" value:"scheduled" ]
- metric.Label = [name:"profile" value:"default-scheduler"  name:"result" value:"unschedulable" ]

metricName = scheduler_pod_scheduling_duration_seconds
- metric.Label = [name:"attempts" value:"1" ]
- metric.Label = [name:"attempts" value:"2" ]
```

Additionally, metrics `*algorithm_predicate*` and `*algorithm_priority*` have been removed, we should update them to `scheduler_framework_extension_point_duration_seconds`:

https://github.com/kubernetes/kubernetes/blob/5bb03db026e796f6c407f796dbf7ddcae0b0eb1f/test/integration/scheduler_perf/scheduler_perf_test.go#L52-L53

/sig scheduling

### PR DESCRIPTION
**Pull request by Huang-Wei**
#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

A HistogramVec can have more than one Histogram. The old logic returned the 0-indexed Histogram, which may make the benchmark results inaccurate. For example, in a Premmption test, it's expected to aggregated all metrics with different `attempts`; instead of return the 0-indexed (which indicates "attempts=1"):

```
metricName = scheduler_pod_scheduling_duration_seconds
- metric.Label = [name:"attempts" value:"1" ]
- metric.Label = [name:"attempts" value:"2" ]
```

This PR tries to aggregate the buckets of all Histograms and return accurate metrics.

Also, in a follow-up PR, we can migrate the old metrics `scheduler_scheduling_algorithm_predicate_evaluation_seconds` and `scheduler_scheduling_algorithm_priority_evaluation_seconds` to `scheduler_framework_extension_point_duration_seconds[extension_point="Filter|Score"].

#### Which issue(s) this PR fixes:

Fixes #99182 #96354

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by Huang-Wei** [comment]
/hold

@ingvagabund Could you review to see if this PR makes sense?

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (ingvagabund) [inline staging/src/k8s.io/component-base/metrics/testutil/metrics.go:180]: Can you add this check to `func (vec HistogramVec) Validate() error`?

File: test/integration/scheduler_perf/util.go
Comment (ingvagabund) [inline test/integration/scheduler_perf/util.go:188]: With vector of histograms, one can get an empty vector. Which will produce NaN for quantiles. Though, when there's one histogram with no buckets, one gets NaN anyway.

**Review by ingvagabund — CHANGES_REQUESTED** [review]
Looks good in overall, just one nit wrt. validation.

File: test/integration/scheduler_perf/util.go
Comment (Huang-Wei) [reply test/integration/scheduler_perf/util.go:188]: How about avoid adding nil Histogram? (although I guess it won't be generated in Promotheus side, but just for double check)

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (Huang-Wei) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:180]: done.

File: test/integration/scheduler_perf/util.go
Comment (ingvagabund) [reply test/integration/scheduler_perf/util.go:188]: Sounds good. I see you already added the check for it as well. Awesome!!!

**Comment by ingvagabund** [comment]
LGTM
Feel free to squash

**Comment by Huang-Wei** [comment]
@logicalhan Could you help review this?

BTW: The changes introduced in this PR (`GetHistogramVecFromGatherer`) only impact scheduler performance test: https://cs.k8s.io/?q=GetHistogramFromGatherer&i=nope&files=&excludeFiles=vendor&repos=

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (adtac) [inline staging/src/k8s.io/component-base/metrics/testutil/metrics.go:289]:
```suggestion
		if LabelsMatch(metric, lvMap) {
```

AFAICT it's okay to pass an empty lvMap to LabelsMatch, it will return true

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (adtac) [inline staging/src/k8s.io/component-base/metrics/testutil/metrics.go:208]: so let's say the first histogram has buckets different from the second histogram such that the second hist's first bucket is smaller than the first hist's first bucket. then the second histogram's quantisation will be incorrectly counted in the cumulative buckets right?

or do we know that all Histograms in a HistogramVc have the same buckets? if so, I think checking for `i == 0` is easier to understand

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (adtac) [inline staging/src/k8s.io/component-base/metrics/testutil/metrics.go:221]: I assume you meant "rest of the samples". If so, wouldn't `count: float64(vec.GetSampleCount())` include ALL of the samples?

File: test/integration/scheduler_perf/util.go
Comment (adtac) [inline test/integration/scheduler_perf/util.go:182]:
```suggestion
func collectHistogramVec(metric string, labels map[string]string) *DataItem {
```

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (Huang-Wei) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:289]: Updated.

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (Huang-Wei) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:208]: We assume all Histograms in a HistogramVc have the same buckets.

> I think checking for i == 0 is easier to understand

Updated.

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (Huang-Wei) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go]: Yes, it's intentional to include all samples. Its previous buckets aggregate bucket from each histogram, and hence the same to this last bucket.

File: test/integration/scheduler_perf/util.go
Comment (Huang-Wei) [reply test/integration/scheduler_perf/util.go]: done.

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (adtac) [inline staging/src/k8s.io/component-base/metrics/testutil/metrics.go:237]: should Clear just clear the internal histograms or should it go back to an empty array? For example, let's say there's a HistogramVec with {attempts=1, attempts=2} as the two histograms.  After a test, we want to Clear() it, but if we only clear the internal histograms and don't empty the slice (`*vec = nil` with `func (vec *HistogramVec) Clear()` as the function signature), we'll still have the same two labels. WDYT?

Comment (Huang-Wei) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:237]: Clear() doesn't do what we expect here. I opened #99472 fixing it.

**Comment by Huang-Wei** [comment]
/remove-sig api-machinery cluster-lifecycle

Kindly ping @logicalhan

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (Huang-Wei) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:237]: @adtac I guess that _may_ be a problem of current logic, so once we switch to #99472 - which calls `legacyregistry.Reset()`:

https://github.com/kubernetes/kubernetes/blob/0e32c43e07dfa9f03f7695ad892bf8235f2e33f1/staging/src/k8s.io/component-base/metrics/registry.go#L305-L312

The Reset() will (IIUC) do the things you mentioned - call the underlying dto.Histogram to reset:

```go
func (m *Histogram) Reset()         { *m = Histogram{} }
```

Let me know if it is what you meant.

**Comment by Huang-Wei** [comment]
@alculquicondor @adtac @ingvagabund Any other comments before I pursue approval from sig-instrumentation?

**Comment by alculquicondor** [comment]
none from me, but I'm unfamiliar with this infrastructure.

**Comment by chendave** [comment]
> A HistogramVec can have more than one Histogram. The old logic returned the 0-indexed Histogram, which may make the benchmark results inaccurate. For example, in a Premmption test, it's expected to aggregated all metrics with different attempts; instead of return the 0-indexed (which indicates "attempts=1"):

Yes, this is what I want to address in this PR: https://github.com/kubernetes/kubernetes/pull/96258/files as well, feel free to move on, I will rebase on this once this got merged.

**Comment by chendave** [comment]
btw, might be great to update the commit to indicate this fix this issue as well, https://github.com/kubernetes/kubernetes/issues/96354

**Comment by ingvagabund** [comment]
None here

**Comment by Huang-Wei** [comment]
> btw, might be great to update the commit message to indicate this fix this issue as well, #96354

Done.

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (adtac) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:237]: No, what I meant was: while the Reset() function should reset each histogram, if there were 2 preemption attempts, we'll still be left with two empty histograms within the HistogramVec. If the next test has only one preemption attempt, we'll still have 2 labels (the second histogram will be empty, but still present). I'm proposing that we *clear* the HistogramVec array itself instead of clearing each individual underlying Histogram.

Comment (Huang-Wei) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:237]: > if there were 2 preemption attempts, we'll still be left with two empty histograms within the HistogramVec.

Not quite. Each test/workload's collector gets the HistogramVec built from scratch, and previous histograms won't be carried over.

I confirmed by building a test:

- put Preemption testcase ahead of Basic
- run a test with a regex matching Preemption and Basic: `-bench=BenchmarkPerfScheduling/\(Preemption\|SchedulingBasic\)/500Nodes`
- then I got the following result
    ```
    # Preemption testcase
    [DEBUG] metricName = scheduler_e2e_scheduling_duration_seconds
    - [DEBUG] metric.label = [name:"profile" value:"default-scheduler"  name:"result" value:"scheduled" ]
    - [DEBUG] metric.label = [name:"profile" value:"default-scheduler"  name:"result" value:"unschedulable" ]
    [DEBUG] metricName = scheduler_pod_scheduling_duration_seconds
    - [DEBUG] metric.label = [name:"attempts" value:"2" ] <== metric "attempts=1" is not recorded, which is good
    
    # Basic testcase
    [DEBUG] metricName = scheduler_e2e_scheduling_duration_seconds
    - [DEBUG] metric.label = [name:"profile" value:"default-scheduler"  name:"result" value:"scheduled" ] <== makes sense, as basic test doesn't have unschedulable pods
    [DEBUG] metricName = scheduler_pod_scheduling_duration_seconds
    - [DEBUG] metric.label = [name:"attempts" value:"1" ] <== makes sense, metric "attempts=2" is not carried over
    ```

**Comment by Huang-Wei** [comment]
Rebased.

/remove-sig api-machinery
/remove-sig cluster-lifecycle

Kindly ping @logicalhan .

**Comment by Huang-Wei** [comment]
/milestone v1.21
as it improves the correctness of scheduler integration benchmark test.

**Comment by palnabarun** [comment]
@Huang-Wei -- is this a change that is critical for v1.21 or can block the release? At this point of the release cycle, we are only considering changes that are release critical.

/milestone clear

(clearing the milestone pending answers to the above)

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (adtac) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go]: Thank you for confirming this, you can resolve this now.

**Comment by adtac** [comment]
/lgtm for scheduler_perf changes

>is this a change that is critical for v1.21 or can block the release?

@palnabarun at the risk of answering for @Huang-Wei, I should mention that this change improves existing tests and we're still within the test freeze; it should not block the release. @Huang-Wei for the final say of course.

**Comment by Huang-Wei** [comment]
Kindly ping @logicalhan .

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (chendave) [inline staging/src/k8s.io/component-base/metrics/testutil/metrics.go:284]: This looks like is problematic when processing the metric like `scheduler_framework_extension_point_duration_seconds`, it will aggregate all the `Histogram` which belong to the family, but what we want is the individual `Histogram` for each extension point, aggregate "filter" and "score" etc. into one vector and do the calculation is meaningless IMHO. 

The issue could be workaround by defining `lvMap` for each of the extension point, but this is not efficient though, it will loop N times for N extension points.

There are two types of `HistogramVec`, something like `attempts` could be done in this way, but the extension points could not be aggregate in this way, we'd better to differentiate those two cases and come up a `HistogramVec` with labels.

**Comment by alculquicondor** [comment]
@ehashman perhaps you can help us reviewing this?

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (chendave) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:284]: /lgtm 

let's tackle this in a follow up review.

**Comment by Huang-Wei** [comment]
@ehashman @logicalhan would you please take a look at this PR? Its motivation was explained in the PR description.

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (logicalhan) [inline staging/src/k8s.io/component-base/metrics/testutil/metrics.go:183]: histogram

Comment (logicalhan) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:183]: I'd call this GetAggregatedSampleCount.

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go
Comment (logicalhan) [inline staging/src/k8s.io/component-base/metrics/testutil/metrics.go:193]: same.

**Review by logicalhan — COMMENTED** [review]
These methods are kind of begging for unit tests..

**Comment by Huang-Wei** [comment]
Thanks @logicalhan! Will add unit tests.

/hold

**Comment by Huang-Wei** [comment]
/retest

Kindly ping @logicalhan .

**Review by logicalhan — COMMENTED** [review]
/lgtm
/approve

**Comment by Huang-Wei** [comment]
Squashed the commits so @chendave can easily rebase to work on a follow-up item.

/milestone v1.23

**Comment by chendave** [comment]
> Squashed the commits so @chendave can easily rebase to work on a follow-up item.

thanks!

/lgtm

**Comment by alculquicondor** [comment]
/retest

@Huang-Wei I believe we can remove the hold?

**Comment by Huang-Wei** [comment]
> @Huang-Wei I believe we can remove the hold?

That's right.

/hold cancel

**Pull request merged**
